### DISCUSSION TITLE
what files other than .xonshrc and .config/xonsh/rc.d/* get loaded on startup?

### DISCUSSION THREAD

**Discussion by petergaultney** [xonsh/xonsh#5672] [Q&A]
i have a startup hang that doesn't appear to be related to any of the files I know about. 

I have checked this using debug prints, but also by doing `xonsh --no-rc` from a working shell, and then manually sourcing all of the files (`source .xonshrc && source .config/xonsh/rc.d/*`) - and the shell works.

But if I just start xonsh directly, I get stuck here, with no prompt.

<img width="570" alt="Screenshot 2024-08-22 at 10 27 27 AM" src="https://github.com/user-attachments/assets/eb6b66bc-d99b-4d44-a2bd-08d4a7bb0e0d">

obviously _something_ is trying to load and is getting stuck, but I can't figure out what it is. Any hints?

Oh, and somewhat relevant: I copied all 4 of these files from a working xonsh config that is the exact same version (0.18.3) on a different machine with the same OS and architecture, and the same Python version.

**Comment by petergaultney** [comment]
a follow up here - I think I've narrowed it down to the `direnv` xontrib, but I don't understand the mechanism of action.

If I disable the line in my .xonshrc that says `xontrib load direnv`, then my shell has no issues starting up with all of its config.

If I _don't_ disable that, and instead load xonsh with `--no-rc` and then manually source the _exact same file_ - it is also fine?

**Reply by anki-code** [reply]
1. As I see in [direnv](https://github.com/74th/xonsh-direnv/blob/f96ecc0670b8744cc5ff8a0a900efe5de1dacb7e/xontrib/direnv.xsh) there is `on_post_rc` event registration. In case of loading the xontrib from xonshrc this event will be fired after loading xonshrc and may be this is the issue. 

2. Also I see global change `$UPDATE_OS_ENVIRON = True`. This could be a bad pattern because can affect on any other xontribs or other experience. Instead of this the right way is to modify env directly or wrap env changes into `with __xonsh__.env.swap(UPDATE_OS_ENVIRON=True): ...`. I mean https://github.com/74th/xonsh-direnv/pull/21

**Reply by petergaultney** [reply]
thanks for looking into this.

to add a bit of extra detail here:

when adding debug prints, i did see all of my rc.d files load successfully. Does the `on_post_rc` event run after all of those are loaded, or only after `.xonshrc` is loaded?

Is there some better way to debug events, etc. if i'm having issues with xonsh startup in the future? i really just wanted a way to do something simple like hit Ctrl-C and have it show me a stack trace of where it was stuck, but I'm sure there are good reasons why a KeyboardInterrupt shouldn't be able to fire during that stage of startup.

maybe some kind of `-vvv` verbose logging that I could turn on?

**Reply by anki-code** [reply]
The best way to debugging is to [use IDE](https://github.com/anki-code/xonsh-developer-toolkit/?tab=readme-ov-file#setup-ide) with breakpoints. It's pretty easy to run xonsh in IDE PyCharm CE (free to use).

It will be cool to have tracing for events. Try to set `XONSH_DEBUG` to 1 or 2 and maybe the information is there. I don't remember.

**Comment by bestlem** [comment]
direnv will load according to .envrc files in the current directory. Where are your .envrc files?

It should also unload when leaving that directory so @anki_code"s point 2 re global environment should not be a problem - this is what direnv does.

**Reply by petergaultney** [reply]
i don't have any .envrc files on this entire computer - it's brand new, and i'm just getting everything set up on it. 

That does seem like a very reasonable question to ask, though. :) is it possible `direnv` might be trying to load... something _else_?  Maybe xonsh is not in my home directory by default when it starts up?

**Reply by bestlem** [reply]
I am running on macOS and I managed to get direnv working sort of.  However I have found xonsh hangs often - not with direnv but all sorts of other things - see errors with [ExecAlias](https://github.com/xonsh/xonsh/pull/5645)..
Changes are being made and I was waiting for those to be released before trying again.


What is the command you are using in Terminal.app settings?



Looking at my xonsh files I see I have
```
if $SHLVL == 1:
  xontrib load direnv
```

If that check is not there then I get the hang as you show when I run it from another shell.
